File: Byte&Buy.AngularSPA/src/app/core/dto/offers/common/user-panel-offer-response.ts
import { Guid } from "guid-typescript";
import { ImageResponse } from "../../image/image-response";
import { OfferStatus } from "../enum/offer-status";

export interface UserPanelOfferResponse{
    id: Guid;
    title: string;
    image: ImageResponse;
    dateCreated: Date;
    dateEdited?: Date;
    status: OfferStatus;
    quantityAvaliable: number;
    type: 'sale' | 'rent';
}

File: Byte&Buy.Backend/src/ByteBuy.Core/DTO/Public/Offer/Common/UserPanelOfferResponse.cs
using ByteBuy.Core.DTO.Public.Image;
using ByteBuy.Core.DTO.Public.Money;
using System.Text.Json.Serialization;

namespace ByteBuy.Core.DTO.Public.Offer.Common;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(UserSalePanelResponse), "sale")]
[JsonDerivedType(typeof(UserRentPanelResponse), "rent")]
public record UserPanelOfferResponse
{
    public Guid Id { get; init; }
    public string Title { get; init; } = null!;
    public ImageResponse Image { get; init; } = null!;
    public DateTime DateCreated { get; init; }
    public DateTime? DateEdited { get; init; }
    public int QuantityAvaliable { get; init; }
}


public record UserSalePanelResponse : UserPanelOfferResponse
{
    public MoneyDto PricePerItem { get; init; } = null!;
}

public record UserRentPanelResponse : UserPanelOfferResponse
{
    public MoneyDto PricePerDay { get; init; } = null!;
    public int MaxRentalDays { get; set; }
}
